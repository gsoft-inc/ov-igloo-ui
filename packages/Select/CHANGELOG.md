# @igloo-ui/select

## 1.1.1

### Patch Changes

- 08a062e: update packages
- Updated dependencies [08a062e]
  - @igloo-ui/list@0.2.1

## 1.1.0

### Minor Changes

- 4ff1942: Added new optional props to List called 'disableTabbing' and 'onOptionBlur'. Also fixed some styling for the List component.

### Patch Changes

- Updated dependencies [4ff1942]
- Updated dependencies [0b8a62f]
  - @igloo-ui/list@0.2.0
  - @igloo-ui/dropdown@1.2.0

## 1.0.7

### Patch Changes

- 84e2d78: Adding a hover effect
- Updated dependencies [84e2d78]
  - @igloo-ui/input@2.0.5

## 1.0.6

### Patch Changes

- 412dbd5: Added the missing onClose event for the Dropdown component inside the Select.

## 1.0.5

### Patch Changes

- a41e1d4: Fix: moved destructuring of props to function parameters so that storybook will automatically show the default value in the props table. It was done in Modal because it needed to send props to useDialog. In this case, @default annotation was used in comments.
- Updated dependencies [a41e1d4]
  - @igloo-ui/dropdown@1.1.4
  - @igloo-ui/icon-button@1.0.1
  - @igloo-ui/input@2.0.4
  - @igloo-ui/list@0.1.3

## 1.0.4

### Patch Changes

- bf478b5: Added missing tests. Fixed missing code in stories when clicking 'Show code'. Updated Dropdown mock. Added new optional prop highlightToday to help with test snapshots.
- Updated dependencies [bf478b5]
  - @igloo-ui/dropdown@1.1.3

## 1.0.3

### Patch Changes

- 55d12e3: fix error with empty package @shared/components
- Updated dependencies [55d12e3]
  - @igloo-ui/list@0.1.2

## 1.0.2

### Patch Changes

- 3f50827: Made dropdown portal to the end of the document so that it works in Modals. Removed unused line in Modal.
- Updated dependencies [3f50827]
  - @igloo-ui/dropdown@1.1.2

## 1.0.1

### Patch Changes

- d5c91f5: link shared components with aliases
- Updated dependencies [e47ae88]
- Updated dependencies [d5c91f5]
  - @igloo-ui/icon-button@1.0.0
  - @igloo-ui/list@0.1.1

## 1.0.0

### Major Changes

- 7dcba93: Initial release of List and Combobox components. Select was updated to use the Dropdown component that was created previously as well as the new List component.

  ## BREAKING CHANGE

  Select types were updated. List component is used inside Select so SelectOption.tsx was removed. Some functionality was changed. When selecting an element in the select using the keyboard, it now just focuses the element, you need to actually hit enter to select it.

### Patch Changes

- Updated dependencies [7dcba93]
  - @igloo-ui/list@0.1.0
  - @igloo-ui/dropdown@1.1.1

## 0.1.4

### Patch Changes

- 6ee9c45: Updated anything causing errors or warnings in the browser console

## 0.1.3

### Patch Changes

- 0640409: Update @igloo-ui/icons to 1.6.0
- 0640409: udpate classnames to 2.3.2

## 0.1.2

### Patch Changes

- d971143: All tests were updated to use testing library and not enzyme

## 0.1.1

### Patch Changes

- 9842a0d: Add dataTest for the Dropdown

## 0.1.0

### Minor Changes

- 39bc518: Feat: Initial release of the select component
