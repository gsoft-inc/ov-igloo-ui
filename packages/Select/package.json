{
    "name": "@igloo-ui/select",
    "version": "1.4.3",
    "main": "dist/Select.js",
    "module": "dist/Select.js",
    "types": "dist/Select.d.ts",
    "styles": "dist/select.css",
    "publishConfig": {
        "access": "public"
    },
    "repository": "git+https://github.com/gsoft-inc/ov-igloo-ui.git",
    "author": "Officevibe Inc.",
    "contributors": [
        "alexandre.beaudoin@gsoft.com"
    ],
    "license": "Apache-2.0",
    "peerDependencies": {
        "react": ">= 16.8.6"
    },
    "files": [
        "dist"
    ],
    "scripts": {
        "build": "rollup -c rollup.config.js"
    },
    "dependencies": {
        "@igloo-ui/dropdown": "^1.8.0",
        "@igloo-ui/list": "^0.6.1",
        "@igloo-ui/icons": "^1.12.2",
        "@igloo-ui/icon-button": "^1.2.0",
        "@igloo-ui/input": "^2.2.1",
        "@igloo-ui/tokens": "^2.0.0",
        "@hopper-ui/tokens": "^3.1.3",
        "@hopper-ui/icons-react16": "1.2.0",
        "classnames": "^2.3.2"
    },
    "devDependencies": {
        "@shared/components": "0.2.1"
    },
    "browserslist": [
        "> 1%",
        "last 2 versions",
        "IE >= 11"
    ]
}
