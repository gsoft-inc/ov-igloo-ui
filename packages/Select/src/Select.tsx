import * as React from 'react';
import cx from 'classnames';

import './select.scss';

import Dropdown from './Dropdown';

import SelectHeader from './SelectHeader';
import SelectOptionComponent from './SelectOption';
import SelectValue from './SelectValue';

export interface SelectOption {
  /** True if the current option can't be selected. */
  disabled?: boolean;
  /** Icon displayed at the front of the option label. */
  icon?: React.ReactNode;
  /** The option label. */
  label: React.ReactNode;
  /** The option value. */
  value: string | number;
}

export interface SelectProps {
  /** Default value displayed in the Select. */
  children?: React.ReactNode;
  /** Add a specific class to the Select. */
  className?: string;
  /** Add a data-test tag for automated tests. */
  dataTest?: string;
  /** Disabled the Select, the user cannot click on it. */
  disabled?: boolean;
  /** The Select is in an error state. */
  error?: boolean;
  /** True for a compact appearance. */
  isCompact?: boolean;
  /** True if the option list is displayed. */
  isOpen?: boolean;
  /** Callback when selected content change. */
  onChange?: (option: SelectOption | any) => void;
  /** List of available options. */
  options: SelectOption[];
  /** The selected option at the initialisation. */
  selectedOption?: SelectOption;
}

const Select: React.FunctionComponent<SelectProps> = (props: SelectProps) => {
  const {
    children,
    className,
    dataTest,
    disabled = false,
    error,
    isCompact = false,
    isOpen = false,
    onChange,
    options,
    selectedOption,
    ...rest
  } = props;

  const selectRef = React.useRef<HTMLDivElement>(null);
  const [currentSelectedOption, setCurrentSelectedOption] =
    React.useState(selectedOption);
  const [showMenu, setShowMenu] = React.useState(isOpen);

  const handleOnKeyDown = (keyboardEvent: {
    key: string;
    code: string;
  }): void => {
    if (keyboardEvent.key === 'Escape' || keyboardEvent.key === 'Esc') {
      if (showMenu) {
        setShowMenu(false);
      }
    }

    if (keyboardEvent.code === 'Space' || keyboardEvent.key === 'Enter') {
      const keepFocus = showMenu;

      setShowMenu(!showMenu);
      if (keepFocus) {
        if (selectRef && selectRef.current) {
          selectRef.current.focus();
        }
      }
    }
  };

  const handleOnClick = (): void => {
    if (disabled) {
      return;
    }

    setShowMenu(!showMenu);
  };

  const generateOptions = (): React.ReactNode => {
    if (!options || options.length === 0) {
      return <Dropdown />;
    }

    const selectOptions = options.map((option: SelectOption) => {
      if (!option) {
        return null;
      }

      const isSelected = option.value === currentSelectedOption?.value ?? false;
      const optionOnClickHandler = (): void => {
        const hasChanged = currentSelectedOption !== option;

        setCurrentSelectedOption(option);

        if (onChange && hasChanged) {
          onChange(currentSelectedOption);
        }
      };

      return (
        <SelectOptionComponent
          disabled={option.disabled}
          icon={option.icon}
          isCompact={isCompact}
          label={option.label}
          onClick={optionOnClickHandler}
          selected={isSelected}
        />
      );
    });

    return <Dropdown isOpen={isOpen}>{selectOptions}</Dropdown>;
  };

  const canShowMenu = showMenu && !disabled;

  const selectClassname = cx('ids-select', className, {
    'ids-select--active': canShowMenu,
    'ids-select--compact': isCompact,
    'ids-select--disabled': disabled,
    'ids-select--error': error,
  });

  return (
    <div
      ref={selectRef}
      className={selectClassname}
      data-test={dataTest}
      onClick={handleOnClick}
      onKeyDown={handleOnKeyDown}
      role="button"
      tabIndex={0}
      {...rest}
    >
      <SelectHeader isOpen={canShowMenu} disabled={disabled}>
        <SelectValue
          disabled={disabled}
          icon={currentSelectedOption?.icon}
          isCompact={isCompact}
          isPlaceholder={!currentSelectedOption}
          label={currentSelectedOption?.label || children}
        />
      </SelectHeader>
      {canShowMenu && generateOptions()}
    </div>
  );
};

export default Select;
