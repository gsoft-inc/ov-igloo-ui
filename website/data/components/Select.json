[
  [
    {
      "tags": {},
      "filePath": "/Users/franck.gaudin/Devel/Igloo/components/packages/Select/src/Dropdown.tsx",
      "description": "",
      "displayName": "Dropdown",
      "methods": [],
      "props": {
        "children": {
          "defaultValue": null,
          "description": "Default value displayed in the Dropdown.",
          "name": "children",
          "parent": {
            "fileName": "components/packages/Select/src/Dropdown.tsx",
            "name": "DropdownProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Select/src/Dropdown.tsx",
              "name": "DropdownProps"
            },
            {
              "fileName": "components/node_modules/@types/react/index.d.ts",
              "name": "TypeLiteral"
            }
          ],
          "required": false,
          "type": { "name": "ReactNode" }
        },
        "isOpen": {
          "defaultValue": null,
          "description": "True if the Dropdown list is displayed.",
          "name": "isOpen",
          "parent": {
            "fileName": "components/packages/Select/src/Dropdown.tsx",
            "name": "DropdownProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Select/src/Dropdown.tsx",
              "name": "DropdownProps"
            }
          ],
          "required": false,
          "type": { "name": "boolean" }
        },
        "onClose": {
          "defaultValue": null,
          "description": "Callback when the user clicks outside the Dropdown.",
          "name": "onClose",
          "parent": {
            "fileName": "components/packages/Select/src/Dropdown.tsx",
            "name": "DropdownProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Select/src/Dropdown.tsx",
              "name": "DropdownProps"
            }
          ],
          "required": false,
          "type": { "name": "() => void" }
        },
        "position": {
          "defaultValue": null,
          "description": "Position of the Dropdown.",
          "name": "position",
          "parent": {
            "fileName": "components/packages/Select/src/Dropdown.tsx",
            "name": "DropdownProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Select/src/Dropdown.tsx",
              "name": "DropdownProps"
            }
          ],
          "required": false,
          "type": { "name": "Position" }
        }
      }
    }
  ],
  [
    {
      "tags": {},
      "filePath": "/Users/franck.gaudin/Devel/Igloo/components/packages/Select/src/Select.tsx",
      "description": "",
      "displayName": "Select",
      "methods": [],
      "props": {
        "children": {
          "defaultValue": null,
          "description": "Default value displayed in the Select.",
          "name": "children",
          "parent": {
            "fileName": "components/packages/Select/src/Select.tsx",
            "name": "SelectProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Select/src/Select.tsx",
              "name": "SelectProps"
            },
            {
              "fileName": "components/node_modules/@types/react/index.d.ts",
              "name": "TypeLiteral"
            }
          ],
          "required": false,
          "type": { "name": "ReactNode" }
        },
        "className": {
          "defaultValue": null,
          "description": "Add a specific class to the Select.",
          "name": "className",
          "parent": {
            "fileName": "components/packages/Select/src/Select.tsx",
            "name": "SelectProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Select/src/Select.tsx",
              "name": "SelectProps"
            }
          ],
          "required": false,
          "type": { "name": "string" }
        },
        "dataTest": {
          "defaultValue": null,
          "description": "Add a data-test tag for automated tests.",
          "name": "dataTest",
          "parent": {
            "fileName": "components/packages/Select/src/Select.tsx",
            "name": "SelectProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Select/src/Select.tsx",
              "name": "SelectProps"
            }
          ],
          "required": false,
          "type": { "name": "string" }
        },
        "disabled": {
          "defaultValue": null,
          "description": "Disable the Select so the user cannot click on it.",
          "name": "disabled",
          "parent": {
            "fileName": "components/packages/Select/src/Select.tsx",
            "name": "SelectProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Select/src/Select.tsx",
              "name": "SelectProps"
            }
          ],
          "required": false,
          "type": { "name": "boolean" }
        },
        "error": {
          "defaultValue": null,
          "description": "The Select is in an error state.",
          "name": "error",
          "parent": {
            "fileName": "components/packages/Select/src/Select.tsx",
            "name": "SelectProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Select/src/Select.tsx",
              "name": "SelectProps"
            }
          ],
          "required": false,
          "type": { "name": "boolean" }
        },
        "isCompact": {
          "defaultValue": null,
          "description": "True for a compact appearance.",
          "name": "isCompact",
          "parent": {
            "fileName": "components/packages/Select/src/Select.tsx",
            "name": "SelectProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Select/src/Select.tsx",
              "name": "SelectProps"
            }
          ],
          "required": false,
          "type": { "name": "boolean" }
        },
        "isOpen": {
          "defaultValue": null,
          "description": "True if the option list is displayed.",
          "name": "isOpen",
          "parent": {
            "fileName": "components/packages/Select/src/Select.tsx",
            "name": "SelectProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Select/src/Select.tsx",
              "name": "SelectProps"
            }
          ],
          "required": false,
          "type": { "name": "boolean" }
        },
        "onChange": {
          "defaultValue": null,
          "description": "Callback when selected content changes.",
          "name": "onChange",
          "parent": {
            "fileName": "components/packages/Select/src/Select.tsx",
            "name": "SelectProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Select/src/Select.tsx",
              "name": "SelectProps"
            }
          ],
          "required": false,
          "type": { "name": "(option: SelectOption) => void" }
        },
        "options": {
          "defaultValue": null,
          "description": "List of available options.",
          "name": "options",
          "parent": {
            "fileName": "components/packages/Select/src/Select.tsx",
            "name": "SelectProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Select/src/Select.tsx",
              "name": "SelectProps"
            }
          ],
          "required": true,
          "type": { "name": "SelectOption[]" }
        },
        "selectedOption": {
          "defaultValue": null,
          "description": "The initial selected option.",
          "name": "selectedOption",
          "parent": {
            "fileName": "components/packages/Select/src/Select.tsx",
            "name": "SelectProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Select/src/Select.tsx",
              "name": "SelectProps"
            }
          ],
          "required": false,
          "type": { "name": "SelectOption" }
        }
      }
    }
  ],
  [
    {
      "tags": {},
      "filePath": "/Users/franck.gaudin/Devel/Igloo/components/packages/Select/src/SelectInput.tsx",
      "description": "",
      "displayName": "SelectInput",
      "methods": [],
      "props": {
        "children": {
          "defaultValue": null,
          "description": "Default value displayed in the Select Input.",
          "name": "children",
          "parent": {
            "fileName": "components/packages/Select/src/SelectInput.tsx",
            "name": "SelectInputProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Select/src/SelectInput.tsx",
              "name": "SelectInputProps"
            },
            {
              "fileName": "components/node_modules/@types/react/index.d.ts",
              "name": "TypeLiteral"
            }
          ],
          "required": false,
          "type": { "name": "ReactNode" }
        },
        "isOpen": {
          "defaultValue": null,
          "description": "True if the option list is displayed.",
          "name": "isOpen",
          "parent": {
            "fileName": "components/packages/Select/src/SelectInput.tsx",
            "name": "SelectInputProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Select/src/SelectInput.tsx",
              "name": "SelectInputProps"
            }
          ],
          "required": false,
          "type": { "name": "boolean" }
        }
      }
    }
  ],
  [],
  []
]
