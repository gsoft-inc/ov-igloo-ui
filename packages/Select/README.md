# Select

Selects are use for user selection in a folding list.

  <ReferenceLinks />

## Installation

To install `@igloo-ui/select` in your project, you will need to run the following command using [npm](https://www.npmjs.com/):

```bash
npm install @igloo-ui/select
```

If you prefer [Yarn](https://classic.yarnpkg.com/en/), use the following command instead:

```bash
yarn add @igloo-ui/select
```

## Usage

Then to use the component in your code just import it!

```jsx
import Select from '@igloo-ui/select';

<Select
  options={[
    { label: 'Option 1', value: '1' },
    { label: 'Option 2', value: '2' },
  ]}
>
  Place holder
</Select>;
```
