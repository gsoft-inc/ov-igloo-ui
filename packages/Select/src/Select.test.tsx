/**
 * @jest-environment jsdom
 */
import React from 'react';
import { shallow } from 'enzyme';

import Select, { SelectOption } from './Select';

const setup = (props = {}, options: SelectOption[] = []) => {
  options =
    options && options.length > 0
      ? options
      : [
          {
            label: 'Option 1',
            value: '1',
          },
          {
            label: 'Option 2',
            value: '2',
          },
        ];

  const component = shallow(<Select options={options}>Hello world</Select>);
  return component;
};

describe('Select', () => {
  test('It should render without errors', () => {
    const select = setup();
    const wrapper = select.find('.ids-select');
    expect(wrapper.length).toBe(1);
  });

  test('It should render a snapshot', () => {
    const select = setup();
    expect(select).toMatchSnapshot();
  });
});
