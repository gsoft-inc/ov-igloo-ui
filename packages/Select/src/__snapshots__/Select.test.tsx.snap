// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select It should render a snapshot 1`] = `ShallowWrapper {}`;
